### PR TITLE
support migrations between different file ID managers

### LINKED ISSUE #28: Cannot switch file ID managers with same database
## Description

Switching file ID managers from e.g. `LocalFileIdManager` to `ArbitraryFileIdManager` doesn't play well with the database:
```
    HTTPServerRequest(protocol='http', host='127.0.0.1:8888', method='GET', uri='/api/yjs/roomid/Untitled.ipynb?1666790876574', version='HTTP/1.1', remote_ip='127.0.0.1')
    Traceback (most recent call last):
      File "/home/david/mambaforge/envs/jupyterlab/lib/python3.10/site-packages/tornado/web.py", line 1713, in _execute
        result = await result
      File "/home/david/github/davidbrochart/jupyter_server_ydoc/jupyter_server_ydoc/ydoc.py", line 310, in get
        idx = file_id_manager.index(path)
      File "/home/david/github/davidbrochart/jupyter_server_fileid/jupyter_server_fileid/manager.py", line 138, in index
        cursor = self.con.execute("INSERT INTO Files (path) VALUES (?)", (path,))
    sqlite3.IntegrityError: NOT NULL constraint failed: Files.ino
```

## Reproduce

1. Run JupyterLab in collaborative mode and `LocalFileIdManager` installed, and open a file.
2. Run again JupyterLab with an `ArbitraryFileIdManager`, and open the file.
3. See the error above.

## Expected behavior

Should we use a different database when switching file ID managers? If yes, this issue can be closed.
Otherwise, I'm not sure we should have filesystem-specific attributes in the database. I understand it is needed for local filesystems, but maybe we should give it a more abstract name? In the end, it is some kind of identifier. And other file ID managers might not have this notion, so maybe we should allow it to be `NULL`?

### PR DESCRIPTION
**Pull request by dlqqq**
## Description

- Adds two methods to the base File ID manager that all implementations should override:
  - `export_rows()`: a static method that returns a generator yielding a tuple of `[id, path]`
  - `import_rows()`: an instance method that writes a tuple yielded by `export_rows()` to the database
- Both `ArbitraryFIDM` and `LocalFIDM` support migrations between each other by default. Upon initialization, both implementations write their module name and class name to a "database manifest", which allows for subsequent reflection by the other FIDM. If a manifest already exists and it has a different module or class name, we perform a **migration** from the previous FIDM to the current FIDM. The migration strategy is roughly as follows:
  - Read the old FIDM's module and class name from the manifest and store it.
  - Backup the existing database by renaming it with the current timestamp, e.g. `2022-11-09T01:55:32-file_id_manager.db`. Bind this to `backup_db_path`.
  - define `PrevManager = getattr(importlib.import_module(prev_module), prev_classname)`
  - finally, migrate the rows over:
    ```
    for row in PrevManager.export_rows(backup_db_path):
      self.import_row(row)
    ```

The key principle is that the previous FIDM defines how to export the rows in its database, and the current FIDM defines how to import the rows into its database.

## Limitations

- Currently, when switching FIDMs, we rely on the fact that the previous FIDM writes its own manifest to the database file. If custom FIDMs forget to do this, the migration will fail. TODO: maybe move the manifest creation logic into `BaseFileIdManager.__init__()`?

- Requires the previous FIDM implementation to use SQLite and not write some other database file format (e.g. MariaDB, PostgresQL) to the path. I think this is constraint is OK to impose.

## Open questions

- How do we communicate the creation of a backup to the user more clearly? I thought about adding a server log, but this could easily get lost in the noise.

## Related issues

- Closes #28.

### CODE REVIEW

**Comment by dlqqq** [comment]
I will need consensus on whether we're OK with imposing the constraint that all File ID manager implementations use SQLite. It's not clear if a general migration strategy would be possible otherwise. Furthermore, if we impose this constraint, we can merge the duplicate `__init__()` logic in `ArbitraryFIDM` and `LocalFIDM` into `BaseFIDM`, and then perform migrations for all custom FIDMs automatically (as long as custom FIDMs meaningfully implement the import and export methods).

**Comment by dlqqq** [comment]
Interesting edge case when migrating from a local FIDM to arbitrary FIDM on Windows. Essentially, local FIDM knows that this is a local Windows path, and calls `os.path.normpath()` on it to map it to lowercase (since there's no distinction b/w upper and lower case in Windows filesystems).

However, arbitrary FIDM doesn't make any assumptions of the local filesystem and is case sensitive even when running on Windows, and hence is not able to return a relative path since it sees the lower-case content root as being different from the content root.

```py
_______________________ test_migrate_local_to_arbitrary _______________________

fid_db_path = 'C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pytest-of-runneradmin\\pytest-0\\test_migrate_local_to_arbitrar0\\data\\fileidmanager_test.db'
jp_root_dir = WindowsPath('C:/Users/runneradmin/AppData/Local/Temp/pytest-of-runneradmin/pytest-0/test_migrate_local_to_arbitrar0/root_dir')
test_path = 'test_path', test_path_child = 'test_path/child'

    def test_migrate_local_to_arbitrary(fid_db_path, jp_root_dir, test_path, test_path_child):
        local = LocalFileIdManager(db_path=fid_db_path, root_dir=str(jp_root_dir))
        local.con.execute("PRAGMA journal_mode = off")
    
        id_1 = local.index(test_path)
        id_2 = local.index(test_path_child)
        del local
        arbitrary = ArbitraryFileIdManager(db_path=fid_db_path, root_dir=str(jp_root_dir))
    
>       assert arbitrary.get_path(id_1) == test_path
E       AssertionError: assert 'c:/users/runneradmin/appdata/local/temp/pytest-of-runneradmin/pytest-0/test_migrate_local_to_arbitrar0/root_dir/test_path' == 'test_path'
E         - test_path
E         + c:/users/runneradmin/appdata/local/temp/pytest-of-runneradmin/pytest-0/test_migrate_local_to_arbitrar0/root_dir/test_path
```

**Comment by dlqqq** [comment]
I just patched the test to fix the above error. I'm not sure if this edge case is worth handling, given that the motivation is unclear (downgrading on a local filesystem) and that the workaround is as simple as providing a normalized contents root on Windows.

File: jupyter_server_fileid/manager.py [hunk 1]
```diff
@@ -65,6 +67,8 @@ class BaseFileIdManager(ABC, LoggingConfigurable, metaclass=FileIdManagerMeta):
         config=True,
     )
 
+    con: Optional[sqlite3.Connection]
```
Comment (kevin-bates) [inline jupyter_server_fileid/manager.py:70]: I wonder if another base class needs to be injected here that contains SQLite-specifics, along with the "xxx_recursive()" methods.  Abstract base classes should be as minimal as possible where _any_ implementation must apply to _all_ derived classes and its clear that `con` would be an issue.

File: jupyter_server_fileid/manager.py [hunk 2]
```diff
@@ -122,9 +132,114 @@ def _copy_recursive(self, from_path: str, to_path: str, path_mgr: Any = os.path)
...
+        self._write_manifest()
+        self._create_tables()
+
+        # finally, migrate the database from backup_db_path to the current database.
+        PrevManager = getattr(importlib.import_module(prev_module), prev_classname)
+        for row in PrevManager.export_rows(backup_db_path):
+            self.import_row(row)
+
+    @staticmethod
+    def export_rows(db_path: str) -> Generator[Tuple[str, str], None, None]:
```
Comment (kevin-bates) [inline jupyter_server_fileid/manager.py:222]: Migrating a row at a time isn't very performant.  I suggest introducing a configurable _batch size_ and dealing with multiple rows per batch.

File: jupyter_server_fileid/manager.py [hunk 3]
```diff
@@ -122,9 +132,114 @@ def _copy_recursive(self, from_path: str, to_path: str, path_mgr: Any = os.path)
...
+        """Exports the Files table by returning a generator that yields one row
+        at a time, encoded as a two-tuple [id, path].
```
Comment (kevin-bates) [inline jupyter_server_fileid/manager.py:224]: Should the exporting code export all columns, with `id` and `path` being required, and let the importing code decide what has value?  This seems more flexible and may have merit that we can't determine yet.

I would also recommend a JSON structure of sorts, with column names as keys.

File: jupyter_server_fileid/manager.py [hunk 4]
```diff
@@ -122,9 +132,114 @@ def _copy_recursive(self, from_path: str, to_path: str, path_mgr: Any = os.path)
...
+        cursor = con.execute("SELECT id, path FROM Files")
+        row = cursor.fetchone()
+        while row is not None:
+            yield row
+            row = cursor.fetchone()
+
+    @abstractmethod
+    def import_row(self, row: Tuple[str, str]) -> None:
+        """Imports a row yielded from `export_rows()` into the Files table."""
+
```
Comment (kevin-bates) [inline jupyter_server_fileid/manager.py:242]: Can we add a `pass` or `...` for readability?

File: jupyter_server_fileid/manager.py [hunk 5]
```diff
@@ -229,26 +346,43 @@ def _validate_root_dir(self, proposal):
...
+        if self._check_manifest():
+            # ensure tables and indices are created
+            self.log.info("ArbitraryFileIdManager : Creating File ID tables and indices.")
+            self._create_tables()
+        else:
+            # migrate database from old manager to current
+            self.log.info(
+                "ArbitraryFileIdManager : Database from different File ID manager detected. Migrating tables and indices."
+            )
+            self._migrate_database()
```
Comment (kevin-bates) [inline jupyter_server_fileid/manager.py:384]: Could this lead to thrashing?  For example, I use two classes of FIDs (e.g., LFID and AFID) configured on my system, neither of which have bothered to configure the db path (plausible if the LFID created the Files table since its schema is a superset of AFID), then won't the AFID trigger migration when it runs, and vice versa when the LFID starts?

I really think this kind of functionality should be in the user's face and performed by a CLI tool that also does `import` (from a source), `export` (to a destination), and `migration` using both.   With a CLI tool, you can also support _any_ database type or storage type and no need for a `Manifest` table.  In addition, you could support migration between databases of the same type (e.g., SQLite).

If you're too strict about class names in the manifest, you could also run into issues where subclasses are treated as distinct classes, despite the fact that their schemas (and base functionality) are identical.

**Review by kevin-bates — COMMENTED** [review]
Hi @dlqqq.  I like the idea of adding `import` and `export` functionality, but don't think performing implicit migrations, one row at a time, is necessarily a good idea.

File: tests/test_manager.py [hunk 6]
```diff
@@ -91,6 +91,35 @@ def _normalize_path_arbitrary(fid_manager, path):
...
+def test_migrate_local_to_arbitrary(fid_db_path, jp_root_dir, test_path, test_path_child):
+    local = LocalFileIdManager(db_path=fid_db_path, root_dir=str(jp_root_dir))
+    local.con.execute("PRAGMA journal_mode = off")
+
+    id_1 = local.index(test_path)
+    id_2 = local.index(test_path_child)
+    del local
+    # need to normalize the arbitrary content root to produce correct API paths
+    arbitrary = ArbitraryFileIdManager(
+        db_path=fid_db_path, root_dir=str(os.path.normcase(jp_root_dir))
```
Comment (kevin-bates) [inline tests/test_manager.py:116]: Good catch, but I think this should be done in the root-dir validator in AFIM, prior to the current conversion of backslashes to forward slashes, since Windows paths are case-insensitive.

File: jupyter_server_fileid/manager.py (hunk 1 above)
Comment (dlqqq) [reply jupyter_server_fileid/manager.py:70]: Hm, yeah I agree that the base class is getting too bloated now. Putting SQLite-specific attributes and methods into a separate class (e.g. `SqliteMixin`) would allow File ID implementations to not use SQLite or implement migrations.

File: jupyter_server_fileid/manager.py (hunk 2 above)
Comment (dlqqq) [reply jupyter_server_fileid/manager.py:222]: 1. Is there a magic "batch size" that is known to be best for migrating databases? Or is it known that bigger batch sizes are always faster, but you are limited by the memory available on your system?

2. If neither are the case, then users would have to run benchmarks on their own hardware beforehand to get the most value out of this batch size trait. Otherwise, they might configure it to some value that actually makes the migration even slower.

I'm open to setting a default batch size if we can come up with some reasonably justified batch size. However, I can't imagine that users will be migrating their file ID databases frequently enough to warrant a configurable batch size trait. Let's tackle performance concerns later if the need arises and not let them block this PR.

File: jupyter_server_fileid/manager.py (hunk 3 above)
Comment (dlqqq) [reply jupyter_server_fileid/manager.py:224]: That's fair, it allows for more complex logic in `import_rows()` that can be added without us breaking the base class API.

File: tests/test_manager.py (hunk 6 above)
Comment (dlqqq) [reply tests/test_manager.py:116]: Sure, let's call `os.path.normcase()` on the contents root only if we determine it to be a local filesystem path via `os.path.isabs()`. In short the logic would look like:

1. Does nothing on POSIX filesystems (since `os.path.normcase()` does nothing in this case)
2. On Windows, only map to lowercase if the contents root is an absolute local path starting with a drive letter.

File: jupyter_server_fileid/manager.py (hunk 5 above)
Comment (dlqqq) [reply jupyter_server_fileid/manager.py:384]: Yes, the migration is being done automatically at the present moment.

> I really think this kind of functionality should be in the user's face and performed by a CLI tool that also does import (from a source), export (to a destination), and migration using both. With a CLI tool, you can also support any database type or storage type and no need for a Manifest table. In addition, you could support migration between databases of the same type (e.g., SQLite).

Wait, but what would `import` and `export` do? These CLI options don't make sense to me. Importing requires somewhere to export to, and vice versa.

I do agree that migrations should be made explicit and only done via a CLI tool. This allows us to convey the backup database path to the end user more easily. However, I'm wary of the old behavior this would lead to:

1. start server with ArbitraryFIDM
2. stop server and restart with LocalFIDM
3. File ID extension fails to load, but this can get easily lost in the logs
4. all extensions depending on File ID are now broken and we could lose data since we're no longer able to index files

Do we know of a way to stop the server from starting if the server extension fails? That way, we can force the user to either migrate their database or switch back to their old FIDM.

File: jupyter_server_fileid/manager.py (hunk 2 above)
Comment (kevin-bates) [reply jupyter_server_fileid/manager.py:222]: Good questions.  There are always trade-offs and "batch-size tuning" is quite common.

> Let's tackle performance concerns later if the need arises and not let them block this PR.

I'm fine with that except that batch size of any value would impact the API and I think we should nail down the API before too long (i.e., the next release).

Could we inject a `List` type hint into the inputs and outputs along with a batch_size configurable that defaults to 1?

```python
    def export_rows(db_path: str) -> Generator[List[Dict]], None, None]:
       ...
```

```python
    def import_rows(self, row: List[Dict]) -> None:
        ...
```
I changed the `Tuple[str,str]` to `Dict` in preparation for returning more columns (next comment).  One side-effect of that, however, is that all of the column names would be redundant on subsequent items in the list.  One approach would be to let the zero-row be a set of tuples of column names and subsequent items be a Tuple of parallel values.

Another approach might be to let the caller indicate the columns (and their order within the "rows" tuples) via a separate tuple, so you'd then have signatures like:
```python
    def export_rows(db_path: str, columns: Tuple) -> Generator[List[Tuple]], None, None]:
       ...
```

```python
    def import_rows(self, columns: Tuple, rows: List[Tuple]) -> None:
        ...
```

File: jupyter_server_fileid/manager.py (hunk 5 above)
Comment (kevin-bates) [reply jupyter_server_fileid/manager.py:384]: > Wait, but what would import and export do? These CLI options don't make sense to me. Importing requires somewhere to export to, and vice versa.

Correct.  Import and Export need sources and destinations and it would be up to the CLI tool to define the types of sources (FIDMs, csv files, jsons, etc.).  _Migration_ essentially reverses those, where the _source_ is exported from and the _destination_ is imported to and would minimally support FIDM "source types".  I.e., the initial implementation of the tool would take FIDM classes from which instances are instantiated and used.  Support for other types could be added as necessary.

Regarding the failure scenario bullets...
Why does the FileID service fail to load when switching to the LFIDM - is it because the schema from the AFIDM is not sufficient and they're both configured to use the same database file (which strikes me as an administration bug)?
Won't _every_ call to the FID service then fail and wouldn't that only go unnoticed if the calling application ignored exceptions?  I hope callers are robust enough to recognize exceptions, but I might be missing something here.

File: jupyter_server_fileid/manager.py (hunk 2 above)
Comment (dlqqq) [reply jupyter_server_fileid/manager.py:222]: Yeah, great point regarding the fact this would be a breaking API change. Let's do this as soon as possible, even before we determine a good batch size to default to.

> I changed the Tuple[str,str] to Dict in preparation for returning more columns (next comment). One side-effect of that, however, is that all of the column names would be redundant on subsequent items in the list.

I think the overhead of serializing each row to a dictionary is negligible compared to the performance of disk reads. Furthermore, the `import_rows()` method would have to build a dictionary mapping each column name to the index in the tuple, so we're not really saving on much performance. Let's stick to representing each row as a dictionary rather than as a tuple.

File: jupyter_server_fileid/manager.py (hunk 5 above)
Comment (dlqqq) [reply jupyter_server_fileid/manager.py:384]: > Correct. Import and Export need sources and destinations and it would be up to the CLI tool to define the types of sources (FIDMs, csv files, jsons, etc.).

Sorry if I'm not understanding you fully, but it almost sounds like you want two additional CLI options in addition to `jupyter server fileid migrate`:

- `jupyter server fileid import`
- `jupyter server fileid export`

My question is what would these options do differently than `migrate`? It seems importing a file ID database and exporting to a different one is the exact same as migrate.

> Why does the FileID service fail to load when switching to the LFIDM - is it because the schema from the AFIDM is not sufficient and they're both configured to use the same database file (which strikes me as an administration bug)?

Yeah, so it'll fail at runtime because the database schemas are different. While this is partially an administration bug, there is a legitimate use case for migrations regardless. Right now, if you want to switch FIDM implementations, you're forced to use a separate DB file, meaning you just lost all the file IDs you previously indexed. Thus you also broke all extensions that depend on file ID.

> Won't every call to the FID service then fail and wouldn't that only go unnoticed if the calling application ignored exceptions? I hope callers are robust enough to recognize exceptions, but I might be missing something here.

Right, but there remains the issue of how to communicate to the user that "because you switched file ID managers, the best way to fix these errors is to migrate your database". I would prefer the server to just not start if the file ID database hasn't been migrated rather than have the user try and debug this on their own. That way, we force the user to take one of three paths:

1. Revert back to the old File ID manager
2. Perform a migration with the CLI tool
3. Drop the old database

I don't want people to have to scratch their heads and figure out why file ID isn't working after switching implementations.

Comment (kevin-bates) [reply jupyter_server_fileid/manager.py:384]: > Sorry if I'm not understanding you fully, but it almost sounds like you want two additional CLI options

I'm just saying that since _migration_ is the "pipe between export and import" a CLI tool could offer different sources.  What we've been primarily talking about are sources of FIDM instances (since they front the database), but I'm saying that you could also support different _output destinations_ where the CLI can store exported data, and different _input sources_ where the CLI tool can get data for importing.

So I could perform an equivalent migration by exporting from FIDM1 to a file `exported.json` and importing from `exported.json` to FIDM2.  That is, the CLI tool manages the "pipe" since it is the caller of `FIDM1.export_rows()` and `FIDM2.import_rows()`.  This can then provide users the ability to have more choices for how they want to migrate (or _backup_) their FID databases.

One application for this may be that the user needs to perform some transformations on the exported rows (e.g., `exported.json`) prior to their import as part of their "migration plan".

We don't necessarily need to expose other sources and split `migrate` into `export` and `import` right away but should have that kind of thing in mind during implementation so that "the pipe" is extensible such that a user could manually intervene between "export" and "import" as part of their migration.

Comment (kevin-bates) [reply jupyter_server_fileid/manager.py:384]: Thinking a bit more on this, the first non-FIDM source type to support would probably want to be `csv`, since you want the ability to read _rows_ from a file and JSON doesn't really lend itself to that format (that I'm aware of).  `.csv` would also be immediately usable in spreadsheet applications in which the transformations could be performed, etc.

Comment (dlqqq) [reply jupyter_server_fileid/manager.py:384]: I don't think we need to support import/export to some secondary file format in this PR. Let's limit the scope of this PR for just migrating between SQLite DBs. However, these are great ideas we should definitely think about implementing later down the road.

File: jupyter_server_fileid/manager.py (hunk 1 above)
Comment (dlqqq) [reply jupyter_server_fileid/manager.py:70]: This PR is growing too large as it is, let's just refactor this in a separate follow-up PR.

File: tests/test_manager.py (hunk 6 above)
Comment (dlqqq) [reply tests/test_manager.py]: Done. 👍

File: jupyter_server_fileid/manager.py (hunk 4 above)
Comment (dlqqq) [reply jupyter_server_fileid/manager.py:234]: 👍

File: jupyter_server_fileid/manager.py (hunk 3 above)
Comment (dlqqq) [reply jupyter_server_fileid/manager.py]: Done 👍

File: jupyter_server_fileid/manager.py (hunk 2 above)
Comment (dlqqq) [reply jupyter_server_fileid/manager.py]: Done 👍

File: jupyter_server_fileid/manager.py [hunk 7]
```diff
@@ -122,9 +132,106 @@ def _copy_recursive(self, from_path: str, to_path: str, path_mgr: Any = os.path)
...
+        os.rename(self.db_path, backup_db_path)
+
+        # recreate files table and indices
+        self.con = sqlite3.connect(self.db_path)
+        self._write_manifest()
+        self._create_tables()
+
+        # finally, migrate the database from backup_db_path to the current database.
+        PrevManager = getattr(importlib.import_module(prev_module), prev_classname)
+        for row in PrevManager.export_rows(backup_db_path):
```
Comment (kevin-bates) [inline jupyter_server_fileid/manager.py:218]: Seems like `row` should be named _`page`_ or _`batch`_ or something to that effect.  A configurable, defaulting to `1` should also get included, otherwise a code change would be required to increase the page size.

Comment (dlqqq) [reply jupyter_server_fileid/manager.py:218]: We can implement actual batching in a follow-up PR (that also provides benchmarks to justify the new default batch size). For now the batch size is hard-coded to 1.

File: jupyter_server_fileid/manager.py [hunk 8]
```diff
@@ -95,6 +99,9 @@ def _from_normalized_path(self, path: Optional[str]) -> Optional[str]:
     def _move_recursive(self, old_path: str, new_path: str, path_mgr: Any = os.path) -> None:
         """Move all children of a given directory at `old_path` to a new
         directory at `new_path`, delimited by `sep`."""
+        if not hasattr(self, "con") or self.con is None:
```
Comment (kevin-bates) [inline jupyter_server_fileid/manager.py:102]: I don't think the attribute existence check is necessary (here and elsewhere).
```suggestion
        if self.con is None:
```

Comment (dlqqq) [reply jupyter_server_fileid/manager.py:102]: Oh yeah, this got added to prevent mypy from complaining. The real fix is to move all these SQLite-specific methods and attributes into a separate mixin like you had suggested previously. I've created an issue to track this separately: https://github.com/jupyter-server/jupyter_server_fileid/issues/53

Comment (kevin-bates) [reply jupyter_server_fileid/manager.py:102]: > Oh yeah, this got added to prevent mypy from complaining. 

Could you please be more specific regarding your `mypy` comment?  I just removed the `not hasattr()` clauses and `pre-commit` (with `mypy`) ran successfully.  I'm just curious about what you're seeing.  Thanks.  I'm using `mypy-0.991`.

Comment (kevin-bates) [reply jupyter_server_fileid/manager.py:102]: Also, I think the `hasattr()` clause in `__del()__` needs to be extended with `and self.con is not None` prior to calling `self.con.commit()` and `self.con.close()` at L513.
